### PR TITLE
[DEM-1112] Disable updating dac values

### PR DESCRIPTION
**Pull request by sinhaneelam**
Updated apply configuration function for D5a instrument adapter, to compare the configuration values to be same as snapshot else raise an exception.

### CODE REVIEW

File: src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py [hunk 1]
```diff
@@ -46,11 +49,34 @@ def __init__(self, address: str, instrument_name: Optional[str] = None) -> None:
             raise SpanValueError('D5a instrument has span unequal to "4v bi"')
 
     def apply(self, config: PythonJsonStructure) -> None:
-        """ Step values for dacs should be part of configuration."""
-        super().apply(config)
+        """ Does not apply config, except for instrument name, to device, but compares config to device settings.
```
Comment (peendebak) [inline src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: This is not true, it does apply settings of the `step` and `inter_delay`

File: src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py [hunk 2]
```diff
@@ -46,11 +49,34 @@ def __init__(self, address: str, instrument_name: Optional[str] = None) -> None:
...
+
+        device_config = self.read(True)
+
+        for parameter in config:
+            if parameter in self._instrument.parameters and hasattr(self._instrument.parameters[parameter], 'set'):
+                if 'value' in config[parameter]:
+                    self._assert_value_matches(config[parameter]['value'], device_config[parameter]['value'], parameter)
+
+        if 'name' in config:
+            self._instrument.name = config['name']
```
Comment (peendebak) [inline src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: You cannot update the name of a `qcodes.Instrument` by setting it, this will not update the name in the station (and probably other places).  I even think it would be better if the Instrument name was read-only.

Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: If the sequence of operation is load & apply config, then create the station. And thereafter reloading the hardware_kernel picks up the latest changes. If, so then use case would pick the latest names.

File: src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py (hunk 1 above)
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: Updated.

File: src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py (hunk 2 above)
Comment (peendebak) [reply src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: @sinhaneelam Not sure what you are saying here, let's discuss offline. Renaming a qcodes instrument will not work.

Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: @peendebak  sure

File: src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py [hunk 3]
```diff
@@ -61,20 +63,43 @@ def test_apply_config(self):
             self.assertEqual(d5a_module_mock(), d5a_adapter.instrument.d5a)
             self.assertEqual(instrument_name, d5a_adapter.instrument.name)
 
+            mocked_snapshot = {'name': 'd5a', 'parameters': self.mock_config}
+            d5a_adapter.instrument.snapshot = MagicMock(return_value=mocked_snapshot)
```
Comment (qSaevar) [inline src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py]: Why does the snapshot need to be mocked now?

File: src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py [hunk 4]
```diff
@@ -61,20 +63,43 @@ def test_apply_config(self):
...
             self.assertEqual(d5a_adapter.instrument.parameters['dac1'].inter_delay, 0.05)
             self.assertEqual(d5a_adapter.instrument.parameters['dac1'].unit, 'V')
 
-            self.mock_config['dac1']['value'] = None
-            d5a_adapter.apply(self.mock_config)
-            warning_text = 'Some parameter values of d5a are None and will not be set!'
-            logger_mock.warning.assert_called_once_with(warning_text)
+            d5a_adapter.instrument.close()
 
+    def test_apply_config_raises_configuration_mismatch_error(self):
```
Comment (qSaevar) [inline src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py]: This is just a copy paste from the previous test right? Maybe all the duplicate code can be extracted

File: src/qilib/configuration_helper/adapters/__init__.py [hunk 5]
```diff
@@ -1,6 +1,5 @@
-from qilib.configuration_helper import InstrumentAdapterFactory
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/__init__.py]: this import is already present in configuration_helper.__init__.py removed it from here for it was duplicate

File: src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py (hunk 3 above)
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py]: read config returns snapshot.

File: src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py [hunk 6]
```diff
@@ -39,15 +42,14 @@ def setUp(self):
             }
         }
 
-    def test_apply_config(self):
+
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py]: removed

File: src/tests/unittests/configuration_helper/adapters/test_common_instrument_adapter.py [hunk 7]
```diff
@@ -18,5 +18,5 @@ def test_change_name_with_config(self):
         config = PythonJsonStructure(name='new_name')
         adapter.apply(config)
         self.assertEqual('DummyInstrumentAdapter_dummy_address', adapter.name)
-        self.assertEqual('new_name', adapter.instrument.name)
+        self.assertNotEqual('new_name', adapter.instrument.name)
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/adapters/test_common_instrument_adapter.py]: true. removed.

File: src/tests/unittests/configuration_helper/adapters/test_ami430_instrument_adapater.py [hunk 8]
```diff
@@ -1,9 +1,11 @@
 import unittest
 from unittest.mock import patch, MagicMock
 
-from qilib.configuration_helper.adapters.ami430_instrument_adapter import AMI430InstrumentAdapter, ConfigurationError
-from tests.test_data.ami430_snapshot import snapshot, bad_config
+from qilib.configuration_helper.adapters.common_instrument_adapter import ConfigurationError
+from qilib.configuration_helper.adapters.ami430_instrument_adapter import AMI430InstrumentAdapter
 
+from tests.test_data.ami430_snapshot import snapshot, bad_config
+from tests.test_data.dummy_instrument import DummyAMI430Instrument
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/adapters/test_ami430_instrument_adapater.py]: added

File: src/tests/unittests/configuration_helper/adapters/test_ami430_instrument_adapater.py [hunk 9]
```diff
@@ -33,24 +35,28 @@ def test_apply_with_name(self):
             adapter = AMI430InstrumentAdapter(address)
         self.assertEqual('no-good-name', adapter.instrument.name)
         adapter.apply(config)
-        self.assertEqual('a_good_name', adapter.instrument.name)
+        self.assertNotEqual('a_good_name', adapter.instrument.name)
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/adapters/test_ami430_instrument_adapater.py]: true.removed.

File: src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py [hunk 10]
```diff
@@ -48,40 +44,17 @@ def __init__(self, address: str, instrument_name: Optional[str] = None) -> None:
         self.field_variation_tolerance = 0.01
 
     def apply(self, config: PythonJsonStructure) -> None:
-        """ Does not apply config, except for instrument name, to device, but compares config to device settings.
+        """ Does not apply config only  compares config to device settings.
 
         Args:
             config: Containing the instrument configuration.
 
-        Raises:
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py]: Yes, added it to doctsring

File: src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py [hunk 11]
```diff
@@ -48,40 +44,17 @@ def __init__(self, address: str, instrument_name: Optional[str] = None) -> None:
...
+        super().compare_config_on_apply(config, parameter_argument)
 
     def _filter_parameters(self, parameters: PythonJsonStructure) -> PythonJsonStructure:
         for values in parameters.values():
             if 'val_mapping' in values:
                 values.pop('val_mapping')
         return parameters
-
-    def _check_field_value(self, config_value: float, device_value: float) -> None:
-        delta = math.fabs(config_value - device_value)
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py]: Removed

File: src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py (hunk 4 above)
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/adapters/test_d5a_instrument_adapter.py]: Removed the extra bits

File: src/tests/unittests/configuration_helper/adapters/test_ami430_instrument_adapater.py [hunk 12]
```diff
@@ -33,24 +35,28 @@ def test_apply_with_name(self):
...
+        adapter.instrument.close()
 
     def test_field_difference_raises_error(self):
         mock_ami430 = MagicMock()
         mock_ami430.snapshot.return_value = {'name': 'magnet', 'parameters': {'field': {'value': -0.0001205}}}
         with patch('qilib.configuration_helper.adapters.ami430_instrument_adapter.AMI430', return_value=mock_ami430):
             address = '192.168.1.128:7180'
             adapter = AMI430InstrumentAdapter(address)
-
+        adapter._instrument = DummyAMI430Instrument('AMI430')
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/adapters/test_ami430_instrument_adapater.py]: Update the structure

File: src/qilib/configuration_helper/adapters/common_instrument_adapter.py [hunk 13]
```diff
@@ -18,18 +18,24 @@
 OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 """
 import logging
+import  math
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/common_instrument_adapter.py]: removed

**Comment by sinhaneelam** [comment]
> 
> 
> Left some inline comments.
> Also I see that type hints have been going far down hill since last time I reviewed a qilib ticket.
> Can you make sure that modules that you touch/edit do not raise mypy errors? And also, maybe, if you are in a good mood, add mypy to travis file?
> 
> to check type hints with mypy you can use:
> `MYPYPATH=./src mypy --strict --ignore-missing-imports -p qilib`

Made changes for adapter

File: src/tests/unittests/configuration_helper/test_instrument_adapter.py [hunk 14]
```diff
@@ -26,6 +27,9 @@ def apply(self, config: PythonJsonStructure) -> None:
             def _filter_parameters(self, parameters: PythonJsonStructure) -> PythonJsonStructure:
                 pass
 
+            def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str) -> bool:
```
Comment (qSaevar) [inline src/tests/unittests/configuration_helper/test_instrument_adapter.py]: Why did you add this method here?

File: src/tests/unittests/configuration_helper/test_instrument_configuration_visitor.py [hunk 15]
```diff
@@ -16,6 +17,9 @@ def __init__(self, address: str):
     def _filter_parameters(self, parameters: PythonJsonStructure) -> PythonJsonStructure:
         return parameters
 
+    def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str) -> bool:
```
Comment (qSaevar) [inline src/tests/unittests/configuration_helper/test_instrument_configuration_visitor.py]: and here?

File: src/tests/unittests/configuration_helper/test_insturment_configuration.py [hunk 16]
```diff
@@ -34,6 +35,10 @@ def apply(self, config: PythonJsonStructure) -> None:
             def _filter_parameters(self, parameters: PythonJsonStructure) -> PythonJsonStructure:
                 pass
 
+            def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str) -> bool:
```
Comment (qSaevar) [inline src/tests/unittests/configuration_helper/test_insturment_configuration.py]: and here?

File: src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py [hunk 17]
```diff
@@ -46,15 +46,23 @@ def __init__(self, address: str, instrument_name: Optional[str] = None) -> None:
...
+        Configuration Parameter with setter commands are matched for equality, on mismatch
+        error is raised
+
         Args:
             config: Containing the instrument configuration.
-
         """
-        parameter_argument= {'field': self.field_variation_tolerance}
-        super().compare_config_on_apply(config, parameter_argument)
+        super()._config_with_setter_command_mismatch_raises_error(config)
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py]: 1. super() not self() : The way to access parameter is same, so put it in super(), we can do with self() also but it would be duplicate code to access parameter
2. Made it _apply

File: src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+from qilib.configuration_helper import InstrumentAdapter
+from qilib.utils import PythonJsonStructure
+
+
+class ConfigurationError(Exception):
+    """ Error to raise if configuration does not match."""
+
+
+class CommonConfigCheckInstrumentAdapter(InstrumentAdapter, ABC):
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py]: Dopped Check from the name

File: src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+    def _config_with_setter_command_mismatch_raises_error(self, config: PythonJsonStructure) -> None:
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py]: Renamed _apply

File: src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        for parameter in config:
+            if parameter in self._instrument.parameters and hasattr(self._instrument.parameters[parameter], 'set'):
+                result = self._compare_config_values(config[parameter]['value'],
+                                                     device_config[parameter]['value'], parameter)
+                if result:
+                    self._raise_configuration_error(config[parameter]['value'], device_config[parameter]['value'],
+                                                    parameter)
+
+    @abstractmethod
+    def _filter_parameters(self, parameters: PythonJsonStructure) -> PythonJsonStructure:
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py]: Removed

File: src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+            PythonJsonStructure: Contains the instrument snapshot without the instrument
+                                 parameters which are filtered out by this function.
+        """
+
+    @abstractmethod
+    def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str) -> bool:
+        """ Comparison logic for  configuration parameter values.
+
+        This function should be overwritten in the subclasses for each specific instrument.
+        In case for an instrument the configuration can not be applied this method should define the comparison
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/common_config_check_instrument_adapter.py]: Updated comment

File: src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py [hunk 22]
```diff
@@ -17,24 +17,27 @@
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 """
-from typing import Optional
+from typing import Any, Optional
 
 from qcodes.instrument_drivers.QuTech.D5a import D5a
 
-from qilib.configuration_helper.adapters import SpiModuleInstrumentAdapter
+from qilib.configuration_helper.adapters.common_config_check_instrument_adapter import \
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: sure

File: src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py [hunk 23]
```diff
@@ -57,5 +64,8 @@ def apply(self, config: PythonJsonStructure) -> None:
         for dac, values in dac_parameters.items():
             self._instrument[dac].step = values['step']
             self._instrument[dac].inter_delay = values['inter_delay']
+        super()._config_with_setter_command_mismatch_raises_error(config)
 
-        super().compare_config_on_apply(config)
+    def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str) -> bool:
+        del parameter
```
[the remark this answers is not in the document]
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: done

File: src/tests/unittests/configuration_helper/test_instrument_adapter.py (hunk 14 above)
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/test_instrument_adapter.py]: Missed to remove ! removed it

File: src/tests/unittests/configuration_helper/test_instrument_configuration_visitor.py (hunk 15 above)
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/test_instrument_configuration_visitor.py]: removed

File: src/tests/unittests/configuration_helper/test_insturment_configuration.py (hunk 16 above)
Comment (sinhaneelam) [reply src/tests/unittests/configuration_helper/test_insturment_configuration.py]: removed

File: src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py [hunk 24]
```diff
@@ -48,40 +43,25 @@ def __init__(self, address: str, instrument_name: Optional[str] = None) -> None:
...
-        if delta > self.field_variation_tolerance:
-            raise ConfigurationError(
-                "Target value for field does not match device value: {}T != {}T".format(config_value, device_value))
-
-    @staticmethod
-    def _assert_value_matches(config_value: Any, device_value: Any, parameter: str) -> None:
-        if config_value != device_value:
-            raise ConfigurationError(
-                "Configuration for {} does not match: '{}' != '{}'".format(parameter, config_value, device_value))
+    def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str = None) -> bool:
```
Comment (qSaevar) [inline src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py]: type `str` can not be `None` by default. you can use  `Optional[str] = None`

File: src/qilib/configuration_helper/adapters/common_config_instrument_adapter.py [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+        for parameter in config:
+            if parameter in self._instrument.parameters and hasattr(self._instrument.parameters[parameter], 'set'):
+                result = self._compare_config_values(config[parameter]['value'],
+                                                     device_config[parameter]['value'], parameter)
+                if result:
+                    self._raise_configuration_error(config[parameter]['value'], device_config[parameter]['value'],
+                                                    parameter)
+
+    @abstractmethod
+    def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str = None) -> bool:
```
Comment (qSaevar) [inline src/qilib/configuration_helper/adapters/common_config_instrument_adapter.py]: same as above

File: src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py [hunk 26]
```diff
@@ -46,11 +47,23 @@ def __init__(self, address: str, instrument_name: Optional[str] = None) -> None:
...
+        """
         unit = config['dac1']['unit']
         self._instrument.set_dac_unit(unit)
         dac_parameters = {param: values for param, values in config.items() if param[0:3] == 'dac'}
         for dac, values in dac_parameters.items():
             self._instrument[dac].step = values['step']
             self._instrument[dac].inter_delay = values['inter_delay']
+        super().apply(config)
+
+    def _compare_config_values(self, config_value: Any, device_value: Any, parameter: str = None) -> bool:
```
Comment (qSaevar) [inline src/qilib/configuration_helper/adapters/d5a_instrument_adapter.py]: same same

File: src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py (hunk 24 above)
Comment (sinhaneelam) [reply src/qilib/configuration_helper/adapters/ami430_instrument_adapter.py]: done
